rust/src/gildedrose.rs

```diff
@@ -30,55 +31,96 @@ impl GildedRose {
+        if item.sell_in <= 0 {
+            item.quality = 0;
+            item.sell_in -= 1;
+            return;
+        };
+
+        // increase value by 3 with less than 3 days to sell
+        if item.sell_in <= 5 {
+            item.quality = std::cmp::min(item.quality + 3, 50);
+            item.sell_in -= 1;
+            return;
+        }
+
+        // increase value by 2 with less than 10 days to sell
+        if item.sell_in <= 10 {
+            item.quality = std::cmp::min(item.quality + 2, 50);
+            item.sell_in -= 1;
+            return;
+        }
+
+        // the default: increase quality by 1
+        item.quality += 1;
+        item.sell_in -= 1;
+    }
+
+    // method to update quality of standard items
+    fn update_quality_standard(item: &mut Item) {
+        // degrade by 2 when sellin date has arrived
+        if item.sell_in <= 0 {
+            item.quality = std::cmp::max(item.quality - 2, 0); // using std max function to ensure min quality is 0
+            item.sell_in -= 1;
+            return;
+        }
+
+        // degrade by 1 when past sellin date
+        item.quality = std::cmp::max(item.quality - 1, 0); // using std max function to ensure min quality is 0
+        item.sell_in -= 1;
+    }
+
+    // method to update quality of conjured items
+    fn update_quality_conjured(item: &mut Item) {
+        // degrade by 4 when sellin date has arrived
+        if item.sell_in <= 0 {
+            item.quality = std::cmp::max(item.quality - 4, 0); // using std max function to ensure min quality is 0
+            item.sell_in -= 1;
+            return;
+        }
+        // degrade by 2 before sellin date
+        item.quality = std::cmp::max(item.quality - 2, 0); // using std max function to ensure min quality is 0
+        item.sell_in -= 1;
+    }
+
+    // method to update quality of items available at the gilded rose
     pub fn update_quality(&mut self) {
+        println!("updating quality");
         for item in &mut self.items {
-            if item.name != "Aged Brie" && item.name != "Backstage passes to a TAFKAL80ETC concert"
-            {
-                if item.quality > 0 {
-                    if item.name != "Sulfuras, Hand of Ragnaros" {
-                        item.quality = item.quality - 1;
-                    }
-                }
+            // each arm selects item for update depedent on name and uses handler for that type
+            // (really wanted to use a match here but had trouble dereferencing from &item)
+            if item.name == "Sulfuras, Hand of Ragnaros" {
+                return;
+            } else if item.name == "Aged Brie" {
+                GildedRose::update_quality_brie(item);
+                return;
+            } else if item.name == "Backstage passes to a TAFKAL80ETC concert" {
+                GildedRose::update_quality_ticket(item);
+                return;
+            } else if item.name == "Conjured Mana Cake" {
```

Comment:
You could use the contains method to check if the name contains the "Conjured" string, I couldn't do it using match